server/src/controllers/subscriptionController.service.ts

```diff
@@ -0,0 +1,144 @@
+      .flatMap(([ticketRestrictionId, ticketQuantity]) => {
+        const ticketRestriction = ticketRestrictionMap.get(ticketRestrictionId);
+        if (!ticketRestriction) {
+          throw new InvalidInputError(
+              422,
+              `Requested tickets no longer available`,
+          );
+        }
+
+        const eventInstance = eventInstanceMap.get(ticketRestriction?.eventinstanceid_fk);
+
+        if (!eventInstance) {
+          throw new InvalidInputError(
+              422,
+              'Requested tickets no longer available',
+          );
+        } else if ((ticketRestriction.availabletickets -= ticketQuantity) < 0 ||
+        (eventInstance.availableseats -= ticketQuantity) < 0) {
+          throw new InvalidInputError(
+              422,
+              'Requested tickets no longer available',
+          );
+        } else if (
+          subscription.seasonsubscriptiontype.subscriptiontype.previewonly &&
+        !eventInstance.ispreview
+        ) {
+          throw new InvalidInputError(
+              422,
+              'Can not redeem preview only subscription ticket for non preview showing',
+          );
+        }
+
+        return Array(ticketQuantity).fill({
+          ticketitem: {
+            create: {
+              ticketrestrictionid_fk: ticketRestrictionId,
+            },
+          },
+        });
+      });
+
+  return {
+    subscriptionTicketItemsToDelete,
+    subscriptionTicketItemsToCreate,
+    eventInstanceIds: [...eventInstanceMap.keys()],
+  };
+};
+
+export interface SeasonSubscriptionType {
+  price: number;
+  ticketlimit: number;
+  subscriptionlimit: number;
+  subscriptiontypeid_fk: number;
+}
+
+export const validateSeasonSubscriptionType = ({
+  price,
+  ticketlimit,
+  subscriptionlimit,
+}: SeasonSubscriptionType, subscriptionsSold = 0, currentTicketLimit = 0) => {
+  if (isNaN(price) || price < 0) {
+    throw new InvalidInputError(422, `Invalid Price (${price})`);
+  } else if (isNaN(ticketlimit) || ticketlimit < 0) {
+    throw new InvalidInputError(422, `Invalid Ticket Limit (${ticketlimit})`);
+  } else if (subscriptionsSold && ticketlimit < currentTicketLimit) {
+    throw new InvalidInputError(422, `Can not reduce ticket limit if any outstanding subscriptions exist`);
+  } else if (isNaN(subscriptionlimit) || subscriptionlimit < subscriptionsSold) {
+    throw new InvalidInputError(422, `Invalid Subscription Limit (${+subscriptionlimit})`);
+  }
+
+  return {price, ticketlimit, subscriptionlimit};
+};
+
+export const validateSubscriptionType = (subscriptionType: any) => {
+  if (isNaN(+subscriptionType.price) || +subscriptionType.price < 0) {
+    throw new InvalidInputError(422, `Invalid Price (${subscriptionType.price})`);
+  }
+  return {
+    name: validateWithRegex(subscriptionType.name, 'Invalid Name', new RegExp('^.{1,255}$')),
+    description: validateWithRegex(subscriptionType.description, `Invalid Description`, new RegExp('^.{1,255}$')),
```

Comment:
For these 2, it might be a good idea to have the error messages be more like `Description must be 255 characters or less`, since this could be directly shown to an admin.

Reply:
Fixed